cache/pmdk_cache.h

```diff
@@ -0,0 +1,369 @@
+  char* key_data;
+
+  Slice key() const { return Slice(key_data, key_length); }
+
+  // Increase the reference count by 1.
+  void Ref() { refs++; }
+
+  // Just reduce the reference count by 1. Return true if it was last reference.
+  bool Unref() {
+    assert(refs > 0);
+    refs--;
+    return refs == 0;
+  }
+
+  // Return true if there are external refs, false otherwise.
+  bool HasRefs() const { return refs > 0; }
+};
+
+struct PersistentEntry{
+  // persistent fields:
+  po::p<size_t> key_size;
+  po::p<size_t> val_size;
+  po::p<size_t> persist_charge;
+  po::persistent_ptr<char[]> key;
+  po::persistent_ptr<char[]> val;
+  po::p<uint32_t> hash;
+  po::persistent_ptr<PersistentEntry> next_hash = nullptr;
+  po::persistent_ptr<PersistentEntry> next_lru = nullptr;
+  po::persistent_ptr<PersistentEntry> prev_lru = nullptr;
+  // persistent flags:
+  po::p<bool> in_cache;
+
+  // TODO: set era to be current era.
+  size_t era = 0;
+  // transient fields, validated with era number:
+  TransientHandle* trans_handle = nullptr;
+
+  // void Ref(){
+  //   assert(trans_handle);
+  //   trans_handle->Ref();
+  // }
+  bool InCache(){
+    return in_cache;
+  }
+  bool HasRefs(){
+    if (trans_handle == nullptr){
+      return false;
+    } else {
+      return trans_handle->HasRefs();
+    }
+  }
+  void SetInCache(bool x){
+    in_cache = x;
+  }
+};
+
+class PersistentRoot;
+
+class PersistTierHashTable{
+  po::persistent_ptr<po::persistent_ptr<PersistentEntry>[]> list_;
+  po::p<uint32_t> length_;
+  po::p<uint32_t> elems_;
+
+  // pop_ is transient. it gets reset during pmdk cache's construction.
+  // TODO: protect this with era number.
+  po::pool<PersistentRoot>* pop_;
+
+  bool KeyEqual(const char* data, size_t size, po::persistent_ptr<PersistentEntry> entry){
+    if (size != entry->key_size){
+      return false;
+    }
+    return (memcmp(data, entry->key.get(), size) == 0);
+  }
+  po::persistent_ptr<PersistentEntry>* FindPointer(const Slice& key, uint32_t hash){
+    po::persistent_ptr<PersistentEntry>* ptr = &list_[hash & (length_-1)];
+    while(*ptr != nullptr && ((*ptr)->hash != hash) || !KeyEqual(key.data(), key.size(), (*ptr))) {
+      ptr = &(*ptr)->next_hash;
+    }
+    return ptr;
+  }
+
+  void Resize(){
+    uint32_t new_length = 16;
+    while(new_length < elems_ * 1.5){
+      new_length *= 2;
+    }
+    po::persistent_ptr<po::persistent_ptr<PersistentEntry>[]> new_list = 
+      po::make_persistent<po::persistent_ptr<PersistentEntry>[]>(new_length);
+    pop_->memset_persist(new_list.get(), 0, sizeof(new_list[0]) * new_length);
+    uint32_t count = 0;
+    for (uint32_t i = 0; i < length_; i++){
+      po::persistent_ptr<PersistentEntry> h = list_[i];
+      while(h != nullptr){
+        po::persistent_ptr<PersistentEntry> next = h->next_hash;
+        uint32_t hash = h->hash;
+        po::persistent_ptr<PersistentEntry>* ptr = &new_list[hash & (new_length - 1)];
+        h->next_hash = *ptr;
+        *ptr = h;
```

Comment:
Do we need to persist this store?

Reply:
This should already be persisted, as `Resize()` is supposed to always be in a transaction.

Reply:
If so, does that mean the `memset_persist` in line 143 is not necessary either?